File: remove_duplicates_from_sorted_list_ii.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: A ListNode
    @return: A ListNode
    """
    def deleteDuplicates(self, head):
        if not head:
            return None
        dummy = ListNode(-1)
        curr = dummy
        while head and head.next:
            while head.val == head.next.val:
                head = head.next
            curr.next = head
            curr = curr.next
        return dummy.next

File: java/RemoveDuplicatesFromSortedListII.java
public class RemoveDuplicatesFromSortedListII {
    /**
     * @param ListNode head is the head of the linked list
     * @return: ListNode head of the linked list
     */
    public static ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode dummy = new ListNode(0);
        ListNode curr = dummy;

        while (head != null) {
            boolean foundDups = false;
            while (head.next != null && head.next.val == head.val) {
                foundDups = true;
                head = head.next;
            }
            if (foundDups) {
                head = head.next;
                continue;
            }

            curr.next = head;
            curr = curr.next;
            head = head.next;
        }
        curr.next = head;

        return dummy.next;
    }
}
